StreamHPC 2023-10-30 (Dynamic Ordering Groundwork)

Naraenda opened the pull request:
This PR contains the groundwork for dynamic ordering. XORWOW has been changed to use dynamic ordering. There will be a second PR that contains the bulk changes on the other generators.

--- 

This PR includes the following merges, ordered from new to old. Noteable changes are bolded.

- 73341441212acb59a6474fddea07c9f8bd2c2add Refactor `if`-chains in `rocrand.cpp`.
- 8ad8f2c8982cc10fa68a006f0e5bec01768b9cad Various doxygen fixes to resolve warnings.
- b4327f3c696cab15cecfdc9c623a87b9e9e90bc4 Fix narrowing conversion in Threefry distribution
- 35cc0305ed4495505c17ebcc02a8668b7b28adf7, b0a08235db021292a1b1298b890fbc080034e38a, 553b10b420a94fe8c290bf49d17b5e6e432b50b1 **XORWOW dynamic ordering implementation**
- 4a0dbc4af676369b50bb433994baecf52c2a170b **Config generator for dynamic ordering**
- ce68cce14826d8452fb8b37bd18cf6008495c9f8 C++ 2017
- 8ddbc852d832caed2ee571e01749ed88d9201004 **Dynamic ordering changes**

---

We still want to add a minor thing to this PR:
- [ ] Config tuning optimizations.

Naraenda commented:
We noticed some merge conflicts. We're rebasing/resolving conflicts and adding the rest of the generators as well. That way we only have to resolve once. I will force-push and undraft when it's ready.

stanleytsang-amd commented on CHANGELOG.md:
```diff
@@ -10,7 +10,8 @@ would copy internal references to the generators and would lead to double free o
   for them. 
 - Improved MT19937 initialization and generation performance.
-- Removed references to and workarounds for deprecated hcc
+- Removed references to and workarounds for deprecated hcc.
```
Please separate any changelog entries for this PR to ROCm 6.1, not ROCm 6.0.

Naraenda replied:
Fixed in 2fa3c1b

stanleytsang-amd commented:
> We noticed some merge conflicts. We're rebasing/resolving conflicts and adding the rest of the generators as well. That way we only have to resolve once. I will force-push and undraft when it's ready.

Thanks.  If you can get this and fix the changelog done by EOD Tuesday, plus any other work for the other Stream update PRs for PRIM/Thrust/CUB that would be greatly appreciated.  We want to get these merged in by Wednesday so we can begin larger QA testing.

Naraenda commented:
> Thanks. If you can get this and fix the changelog done by EOD Tuesday, plus any other work for the other Stream update PRs for PRIM/Thrust/CUB that would be greatly appreciated. We want to get these merged in by Wednesday so we can begin larger QA testing.

@stanleytsang-amd Apologies for the delay. I've added the latest changes and rebased. The summary of commits has also been updated.

Currently MT19937 is missing dynamic ordering as it requires a bit of restructuring (and [dynamic ordering for MT19937 is not supported in cuRAND](https://docs.nvidia.com/cuda/curand/host-api-overview.html#order)).

stanleytsang-amd commented:
> > Thanks. If you can get this and fix the changelog done by EOD Tuesday, plus any other work for the other Stream update PRs for PRIM/Thrust/CUB that would be greatly appreciated. We want to get these merged in by Wednesday so we can begin larger QA testing.
> 
> @stanleytsang-amd Apologies for the delay. I've added the latest changes and rebased. The summary of commits has also been updated.
> 
> Currently MT19937 is missing dynamic ordering as it requires a bit of restructuring (and [dynamic ordering for MT19937 is not supported in cuRAND](https://docs.nvidia.com/cuda/curand/host-api-overview.html#order)).

No worries.  Currently I'm seeing a build error, but it only shows up on SLES.  I will share with you the docker image with you via email to help you debug:

[ 58%] Building CXX object test/internal/CMakeFiles/test_rocrand_generator_type.dir/test_rocrand_generator_type.cpp.o
/var/jenkins_home/workspace/main_precheckin_rocRAND_PR-404/NYlrYx5ol/rocrand/test/internal/test_cpp_utils.cpp:55:20: error: no viable constructor or deduction guide for deduction of template arguments of 'set'
    const std::set combination_set(combinations.begin(), combinations.end());
                   ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:166:7: note: candidate template ignored: couldn't infer template argument '_Key'
      set(const _Compare& __comp,
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:181:2: note: candidate template ignored: couldn't infer template argument '_Key'
        set(_InputIterator __first, _InputIterator __last)
        ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:233:7: note: candidate template ignored: could not match 'initializer_list<value_type>' (aka 'initializer_list<_Key>') against 'const_iterator' (aka 'const std::array<int, 3> *')
      set(initializer_list<value_type> __l,
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:245:7: note: candidate template ignored: could not match 'set<_Key, _Compare, _Alloc>' against 'const_iterator' (aka 'const std::array<int, 3> *')
      set(const set& __x, const allocator_type& __a)
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:249:7: note: candidate template ignored: could not match 'set<_Key, _Compare, _Alloc>' against 'const_iterator' (aka 'const std::array<int, 3> *')
      set(set&& __x, const allocator_type& __a)
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:255:7: note: candidate template ignored: could not match 'initializer_list<value_type>' (aka 'initializer_list<_Key>') against 'const_iterator' (aka 'const std::array<int, 3> *')
      set(initializer_list<value_type> __l, const allocator_type& __a)
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:213:7: note: candidate function template not viable: requires 1 argument, but 2 were provided
      set(const set&) = default;
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:221:7: note: candidate function template not viable: requires 1 argument, but 2 were provided
      set(set&&) = default;
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:241:7: note: candidate function template not viable: requires single argument '__a', but 2 arguments were provided
      set(const allocator_type& __a)
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:93:11: note: candidate function template not viable: requires 1 argument, but 2 were provided
    class set
          ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:261:2: note: candidate function template not viable: requires 3 arguments, but 2 were provided
        set(_InputIterator __first, _InputIterator __last,
        ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:157:7: note: candidate function template not viable: requires 0 arguments, but 2 were provided
      set() = default;
      ^
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../include/c++/7/bits/stl_set.h:198:2: note: candidate function template not viable: requires at least 3 arguments, but 2 were provided
        set(_InputIterator __first, _InputIterator __last,
        ^
1 error generated when compiling for gfx908.

Naraenda commented:
> No worries. Currently I'm seeing a build error, but it only shows up on SLES. I will share with you the docker image with you via email to help you debug: [...]

There might be a few dependencies missing since this PR bumps the minimum C++ standard to C++17. The build works on opensuse/leap:15.5 (via Docker). I have installed the following GCC-packages.

```plaintext
# zypper search gcc --installed-only | grep i+
i+ | gcc                                  | The system GNU C Compiler                                      | package
i+ | gcc-c++                              | The system GNU C++ Compiler                                    | package
i+ | gcc12                                | The GNU C Compiler and Support Files                           | package
i+ | gcc12-c++                            | The GNU C++ Compiler                                           | package
```

stanleytsang-amd commented:
@Naraenda I've spoken with @lawruble13 and the recommendation is to deal with the SLES + gcc10 issue in the appropriate Jenkins setup, and document the issue accordingly for end-users.  Feel free to make the necessary changes.  Thanks.

Maetveis commented:
> @Naraenda I've spoken with @lawruble13 and the recommendation is to deal with the SLES + gcc10 issue in the appropriate Jenkins setup, and document the issue accordingly for end-users. Feel free to make the necessary changes. Thanks.

For the jenkins build installing a later gcc is required: [would be here in rocJenkins](https://github.com/ROCm/rocJenkins/blob/pong/resources/com/amd/docker/dockerfile-build-sles15.1-cuda11#L28) if I understand the setup correctly.

I would recommend `gcc13-c++` as the latest available gcc version. From my testing no other modification is required, clang picks up the latest available standard library by default. I'll push a commit soon adding this to rocRAND documentation for end users.

EDIT: A more conservative version is `gcc9-c++`, this is the default on Ubuntu 20 (which we use in our CI internally) and available in both RHEL8 and SLES 15.

stanleytsang-amd commented:
> > @Naraenda I've spoken with @lawruble13 and the recommendation is to deal with the SLES + gcc10 issue in the appropriate Jenkins setup, and document the issue accordingly for end-users. Feel free to make the necessary changes. Thanks.
> 
> For the jenkins build installing a later gcc is required: [would be here in rocJenkins](https://github.com/ROCm/rocJenkins/blob/pong/resources/com/amd/docker/dockerfile-build-sles15.1-cuda11#L28) if I understand the setup correctly.
> 
> I would recommend `gcc13-c++` as the latest available gcc version. From my testing no other modification is required, clang picks up the latest available standard library by default. I'll push a commit soon adding this to rocRAND documentation for end users.
> 
> EDIT: A more conservative version is `gcc9-c++`, this is the default on Ubuntu 20 (which we use in our CI internally) and available in both RHEL8 and SLES 15.

Thanks for the reply.  We will fix this in rocJenkins,  then we'll merge this PR in.

stanleytsang-amd commented:
PR to rocJenkins to enable gcc12 submitted last Friday: https://github.com/ROCm/rocJenkins/pull/668

Naraenda commented:
@stanleytsang-amd The latest changes have been pushed. I've updated the PR description with a description of each first parent commit.

CC @mfep @Maetveis

mfep commented:
> @stanleytsang-amd The latest changes have been pushed. I've updated the PR description with a description of each first parent commit.
> 
> CC @mfep @Maetveis

There were a few problems with the previous rebase. Those should be fixed now.

stanleytsang-amd commented:
@Naraenda can you target the develop branch please? Not staging branch.

mfep commented:
Rebased to `develop`.

stanleytsang-amd commented:
All tests pass, excellent.  @mfep Is this PR complete with all the desired changes now?  Or are there still more upstream commits?
Also, can we update the changelog to reflect this will be in for ROCm 6.2 - apologies for the delay in updating rocJenkins, because of that we missed the 6.1 window.

mfep commented:
Thanks, @stanleytsang-amd! I've just pushed another batch of commits, that we hope are the last for the current PR. With these, the implementation, interface and documentation of the dynamic ordering are complete, and only the tuning part for some GPU architectures remain. I'll contact @umfranzw about this shortly.

The version numbers and targeted ROCm release were also updated in the changelog and main CMakeList, as requested.

umfranzw commented:
Thanks @mfep. If it sounds all right to you, I will follow up with a smaller PR later following this one that adds support for the additional GPU architectures.

mfep commented:
Rebased to `develop` again to resolve merge conflicts.

umfranzw commented:
@mfep Apologies for the delay on the review here - Stanley is currently away, but I will try to get to it by Monday at the latest.

mfep commented:
@umfranzw Thank you! Rebased to `develop` to resolve merge conflicts.

umfranzw reviewed (approved):
I think this looks ready to go.

RobsonRLemos reviewed (approved):
LGTM.

mfep commented:
@umfranzw @RobsonRLemos @doctorcolinsmith Thank you for the reviews. For me, it is not clear whose responsibility it is to click the merge button, can we clarify that?

Some Stream HPC employees (not including me) are authorized to merge the PR. Should we do so?

umfranzw commented:
Hi @mfep, I can merge this for now, and follow up once I find out the answer to your question.
